src/DataFramesMeta.jl

```diff
@@ -103,74 +103,134 @@ macro col(kw)
     esc(fun_to_vec(kw))
 end
 
+is_simple_function_call(x) = false
+function is_simple_function_call(expr::Expr)
+    (expr.head == :call
+        && length(expr.args) >= 2
+        && expr.args[1] isa Symbol
+        && all(x -> x isa QuoteNode, expr.args[2:end]))
+end
+
+is_simple_broadcast_call(x) = false
+function is_simple_broadcast_call(expr::Expr)
+    (expr.head == :.
+        && length(expr.args) == 2
+        && expr.args[1] isa Symbol
+        && expr.args[2] isa Expr
+        && expr.args[2].head == :tuple
+        && all(x -> x isa QuoteNode, expr.args[2].args))
+end
+
+
+function broadcast_f_expr(f)
+    # theoretically we could use ByRow everywhere, but maybe we can save some compilation
+    # here and there by using the Base functionality where it's available, as other code
+    # might have already called that
+    if VERSION >= v"1.6.0-beta"
+        :(Base.BroadcastFunction($f))
+    else
+        :(DataFrames.ByRow($f))
+    end
+end
+
+
 # `nolhs` needs to be `true` when we have syntax of the form
 # `@combine(gd, fun(:x, :y))` where `fun` returns a `table` object.
 # We don't create the "new name" pair because new names are given
 # by the table.
 function fun_to_vec(kw::Expr; nolhs::Bool = false, gensym_names::Bool = false)
     # nolhs: f(:x) where f returns a Table
     # !nolhs, y = g(:x)
-    if kw.head === :(=) || kw.head === :kw || nolhs
-        membernames = Dict{Any, Symbol}()
-        if nolhs
-            # act on f(:x)
-            body = replace_syms!(kw, membernames)
-        else
-            # act on g(:x)
-            body = replace_syms!(kw.args[2], membernames)
+
+    if !(kw.head === :(=) || kw.head === :kw || nolhs)
+        throw(ArgumentError("Expressions not of the form `y = f(:x)` currently disallowed."))
+    end
+
+    function_expr = nolhs ? kw : kw.args[2]
+
+    # check cases where we can avoid creating an anonymous function
+
+    # f(:x, ...) into [:x, ...] => f
+    if is_simple_function_call(function_expr)
+        # extract source symbols from quotenodes
+        source = [q.value for q in function_expr.args[2:end]]
+        fun = function_expr.args[1]
+        # some normal-looking calls are actually broadcasts, like
+        # .+ .- etc., if we just pass on those symbols as function calls
+        # we get UndefVarErrors
+        # instead we transform the symbols to their non-broadcast versions
+        # and then use broadcast wrappers
+        if startswith(string(fun), ".")
+            f_sym_without_dot = Symbol(chop(string(fun), head = 1, tail = 0))
+            fun = broadcast_f_expr(f_sym_without_dot)
         end
 
+    # f.(:x, ...) into [:x, ...] => BroadcastFunction(f)
+    elseif is_simple_broadcast_call(function_expr)
+        # extract source symbols from quotenodes
+        source = [q.value for q in function_expr.args[2].args]
```

Comment:
I'm not sure I'm following this. In the other branches, `source` is an expression. Here it's an actual vector?

Reply:
the diff is weird because I un-nested the whole if expression and put the guard at the top. I find it hard to read otherwise

Reply:
hm maybe I misread the other code with `source`, could be that you're right and it should also be an expression, I didn't check what `make_source_concrete` does. Although it's interesting that it still worked ;)